[WebPrivacy] Reduce code duplication in helper classes used to manage remotely updatable lists

whsieh opened the pull request:
#### 617326575536554ba7faf2a1f95788f6714f38c1
<pre>
[WebPrivacy] Reduce code duplication in helper classes used to manage remotely updatable lists
<a href="https://bugs.webkit.org/show_bug.cgi?id=279469">https://bugs.webkit.org/show_bug.cgi?id=279469</a>

Reviewed by Charlie Wolfe and Abrar Rahman Protyasha.

Ahead of making further changes in this area, clean up the following helper classes to consolidate
(currently) duplicated code underneath common base classes. See below for more details.

No change in behavior.

* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
(WebKit::ListDataObserver::create):
(WebKit::ListDataObserver::ListDataObserver):

We currently have 3 nearly-identical classes, that all fulfill the same purpose of wrapping a
callback in a weakly-held C++ object:

- LinkDecorationFilteringDataObserver
- StorageAccessPromptQuirkObserver
- StorageAccessUserAgentStringQuirkObserver

Simply consolidate these three into a single class: `WebKit::ListDataObserver`.

(WebKit::ListDataController::shared):
(WebKit::ListDataController::setCachedListDataForTesting):
(WebKit::ListDataController::cachedListData const):
(WebKit::ListDataController::setCachedListData):
(WebKit::ListDataController::didUpdateCachedListData):

Add two helper classes:

1.  `ListDataControllerBase`, a virtual, non-templated class which implements `observeUpdates()` and
    `initialize()`.

2.  `ListDataController&lt;S, T&gt;`, a templated class which derives from `ListDataControllerBase` and
    exposes getters and setters for the backing `T m_cachedListData`.

This split between non-templated and templated classes allows us to keep the implementations of
`observeUpdates` and `initialize` in the C++ source (rather than an inline header), which is
necessary since the implementation of `observeUpdates` calls into ObjC code but this header is
included in C++ sources.

(WebKit::LinkDecorationFilteringController::cachedListData const):
(WebKit::LinkDecorationFilteringDataObserver::create): Deleted.
(WebKit::LinkDecorationFilteringDataObserver::invokeCallback): Deleted.
(WebKit::LinkDecorationFilteringDataObserver::LinkDecorationFilteringDataObserver): Deleted.
(WebKit::LinkDecorationFilteringController::cachedStrings const): Deleted.
(WebKit::StorageAccessPromptQuirkObserver::create): Deleted.
(WebKit::StorageAccessPromptQuirkObserver::invokeCallback): Deleted.
(WebKit::StorageAccessPromptQuirkObserver::StorageAccessPromptQuirkObserver): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkObserver::create): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkObserver::invokeCallback): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkObserver::StorageAccessUserAgentStringQuirkObserver): Deleted.
(WebKit::StorageAccessPromptQuirkController::cachedQuirks const): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::cachedQuirks const): Deleted.
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener initWithType:callback:]):
(-[WKWebPrivacyNotificationListener dealloc]):
(-[WKWebPrivacyNotificationListener didUpdate:]):

Refactor `WKWebPrivacyNotificationListener` so that it takes a single callback block and resource
type, and only calls the callback if the notification&apos;s resource type matches. This is much cleaner
than enumerating every single list type as a member, but only using a single member for any given
listener instance.

(WebKit::ListDataControllerBase::observeUpdates):
(WebKit::ListDataControllerBase::initialize):
(WebKit::LinkDecorationFilteringController::resourceType const):
(WebKit::LinkDecorationFilteringController::updateList):
(WebKit::requestLinkDecorationFilteringData):
(WebKit::StorageAccessPromptQuirkController::resourceType const):
(WebKit::StorageAccessPromptQuirkController::didUpdateCachedListData):
(WebKit::StorageAccessPromptQuirkController::updateList):
(WebKit::StorageAccessUserAgentStringQuirkController::resourceType const):
(WebKit::StorageAccessUserAgentStringQuirkController::updateList):
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::RestrictedOpenerDomainsController::update):
(WebKit::TrackerAddressLookupInfo::populateIfNeeded):
(WebKit::TrackerDomainLookupInfo::populateIfNeeded):
(WebKit::canUseWebPrivacyFramework): Deleted.
(WebKit::resourceDataChangedNotificationName): Deleted.
(WebKit::notificationUserInfoResourceTypeKey): Deleted.

Remove several workarounds that are no longer necessary, since we dropped support for macOS Monterey
on trunk.

(-[WKWebPrivacyNotificationListener init]): Deleted.
(-[WKWebPrivacyNotificationListener listenForLinkFilteringDataChanges:]): Deleted.
(-[WKWebPrivacyNotificationListener listenForStorageAccessPromptQuirkChanges:]): Deleted.
(-[WKWebPrivacyNotificationListener listenForStorageAccessUserAgentStringQuirkChanges:]): Deleted.
(-[WKWebPrivacyNotificationListener listenForRestrictedOpenerDomainsChanges:]): Deleted.
(WebKit::LinkDecorationFilteringController::shared): Deleted.
(WebKit::LinkDecorationFilteringController::observeUpdates): Deleted.
(WebKit::LinkDecorationFilteringController::setCachedStrings): Deleted.
(WebKit::LinkDecorationFilteringController::updateStrings): Deleted.
(WebKit::StorageAccessPromptQuirkController::shared): Deleted.
(WebKit::StorageAccessPromptQuirkController::observeUpdates): Deleted.
(WebKit::StorageAccessPromptQuirkController::setCachedQuirks): Deleted.
(WebKit::StorageAccessPromptQuirkController::setCachedQuirksForTesting): Deleted.
(WebKit::StorageAccessPromptQuirkController::initialize): Deleted.
(WebKit::StorageAccessPromptQuirkController::updateQuirks): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::shared): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::observeUpdates): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirks): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirksForTesting): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::initialize): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::updateQuirks): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
(WebKit::WebProcessPool::platformInitializeNetworkProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
(WebKit::NetworkProcessProxy::NetworkProcessProxy):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::sendCachedLinkDecorationFilteringData):
(WebKit::WebPageProxy::waitForInitialLinkDecorationFilteringData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setStorageAccessPromptQuirkForTesting):
(WebKit::WebsiteDataStore::setUserAgentStringQuirkForTesting):

Canonical link: <a href="https://commits.webkit.org/283464@main">https://commits.webkit.org/283464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77bbd9b5189346896620c7c3c4ef2a0cd1fdc8c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45730 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70388 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16966 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53529 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17248 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53217 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11829 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69422 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42158 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57446 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33869 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38829 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15841 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60725 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15175 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72091 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10312 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14546 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60542 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10344 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60853 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14627 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8512 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2130 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41537 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42614 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43797 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42357 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -50,128 +53,109 @@ enum class RestrictedOpenerType : uint8_t;
-
-    Ref<StorageAccessPromptQuirkObserver> observeUpdates(Function<void()>&&);
+    void updateList(CompletionHandler<void()>&&) final;
+    const Vector<WebCore::LinkDecorationFilteringData>& cachedListData() const { return m_cachedListData; }
```
You shouldn't need this definition given `const BackingDataType& cachedListData() const { return m_cachedListData; }`

whsieh replied:
Whoops, meant to remove this — good catch!

aprotyas commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -36,6 +36,9 @@
 #if ENABLE(ADVANCED_PRIVACY_PROTECTIONS)
 #import <WebCore/LinkDecorationFilteringData.h>
 #import <WebCore/OrganizationStorageAccessPromptQuirk.h>
+#ifdef __OBJC__
+#import <pal/spi/cocoa/WebPrivacySPI.h>
+#endif
```
Wouldn't we unconditionally have `__OBJC__` under `Platform/cocoa/`?

whsieh replied:
Ah, so the header is actually included in `.cpp` (non-ObjC) files behind Cocoa-specific platform or feature flags.

aprotyas commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -285,44 +219,15 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
-    Ref observer = StorageAccessPromptQuirkObserver::create(WTFMove(completionHandler));
-    m_observers.add(observer.get());
-    return observer;
+    return static_cast<WPResourceType>(WPResourceTypeStorageAccessPromptQuirksData);
```
Is there a need to cast here? Looking at `LinkDecorationFilteringController::resourceType()` where we don't.

whsieh replied:
Unfortunately, I think we need a cast to keep the non-internal builds passing :/ (see the failing builds from earlier)

aprotyas commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -285,44 +219,15 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
-        observer.invokeCallback();
-    });
+    m_cachedListData.shrinkToFit();
+    RELEASE_LOG(ResourceLoadStatistics, "StorageAccessPromptQuirkController::setCachedQuirks: Loaded %lu storage access prompt(s) quirks from WebPrivacy.", m_cachedListData.size());
```
> `"StorageAccessPromptQuirkController::setCachedQuirks:`

Do you still want the logs to say that?

whsieh replied:
Oh, great catch — I should change this as well!

aprotyas reviewed (approved):
LGTM! Some minor comments.

whsieh committed:
[WebPrivacy] Reduce code duplication in helper classes used to manage remotely updatable lists
https://bugs.webkit.org/show_bug.cgi?id=279469

Reviewed by Charlie Wolfe and Abrar Rahman Protyasha.

Ahead of making further changes in this area, clean up the following helper classes to consolidate
(currently) duplicated code underneath common base classes. See below for more details.

No change in behavior.

* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
(WebKit::ListDataObserver::create):
(WebKit::ListDataObserver::ListDataObserver):

We currently have 3 nearly-identical classes, that all fulfill the same purpose of wrapping a
callback in a weakly-held C++ object:

- LinkDecorationFilteringDataObserver
- StorageAccessPromptQuirkObserver
- StorageAccessUserAgentStringQuirkObserver

Simply consolidate these three into a single class: `WebKit::ListDataObserver`.

(WebKit::ListDataController::shared):
(WebKit::ListDataController::setCachedListDataForTesting):
(WebKit::ListDataController::cachedListData const):
(WebKit::ListDataController::setCachedListData):
(WebKit::ListDataController::didUpdateCachedListData):

Add two helper classes:

1.  `ListDataControllerBase`, a virtual, non-templated class which implements `observeUpdates()` and
    `initialize()`.

2.  `ListDataController<S, T>`, a templated class which derives from `ListDataControllerBase` and
    exposes getters and setters for the backing `T m_cachedListData`.

This split between non-templated and templated classes allows us to keep the implementations of
`observeUpdates` and `initialize` in the C++ source (rather than an inline header), which is
necessary since the implementation of `observeUpdates` calls into ObjC code but this header is
included in C++ sources.

(WebKit::LinkDecorationFilteringController::cachedListData const):
(WebKit::LinkDecorationFilteringDataObserver::create): Deleted.
(WebKit::LinkDecorationFilteringDataObserver::invokeCallback): Deleted.
(WebKit::LinkDecorationFilteringDataObserver::LinkDecorationFilteringDataObserver): Deleted.
(WebKit::LinkDecorationFilteringController::cachedStrings const): Deleted.
(WebKit::StorageAccessPromptQuirkObserver::create): Deleted.
(WebKit::StorageAccessPromptQuirkObserver::invokeCallback): Deleted.
(WebKit::StorageAccessPromptQuirkObserver::StorageAccessPromptQuirkObserver): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkObserver::create): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkObserver::invokeCallback): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkObserver::StorageAccessUserAgentStringQuirkObserver): Deleted.
(WebKit::StorageAccessPromptQuirkController::cachedQuirks const): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::cachedQuirks const): Deleted.
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener initWithType:callback:]):
(-[WKWebPrivacyNotificationListener dealloc]):
(-[WKWebPrivacyNotificationListener didUpdate:]):

Refactor `WKWebPrivacyNotificationListener` so that it takes a single callback block and resource
type, and only calls the callback if the notification's resource type matches. This is much cleaner
than enumerating every single list type as a member, but only using a single member for any given
listener instance.

(WebKit::ListDataControllerBase::observeUpdates):
(WebKit::ListDataControllerBase::initialize):
(WebKit::LinkDecorationFilteringController::resourceType const):
(WebKit::LinkDecorationFilteringController::updateList):
(WebKit::requestLinkDecorationFilteringData):
(WebKit::StorageAccessPromptQuirkController::resourceType const):
(WebKit::StorageAccessPromptQuirkController::didUpdateCachedListData):
(WebKit::StorageAccessPromptQuirkController::updateList):
(WebKit::StorageAccessUserAgentStringQuirkController::resourceType const):
(WebKit::StorageAccessUserAgentStringQuirkController::updateList):
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::RestrictedOpenerDomainsController::update):
(WebKit::TrackerAddressLookupInfo::populateIfNeeded):
(WebKit::TrackerDomainLookupInfo::populateIfNeeded):
(WebKit::canUseWebPrivacyFramework): Deleted.
(WebKit::resourceDataChangedNotificationName): Deleted.
(WebKit::notificationUserInfoResourceTypeKey): Deleted.

Remove several workarounds that are no longer necessary, since we dropped support for macOS Monterey
on trunk.

(-[WKWebPrivacyNotificationListener init]): Deleted.
(-[WKWebPrivacyNotificationListener listenForLinkFilteringDataChanges:]): Deleted.
(-[WKWebPrivacyNotificationListener listenForStorageAccessPromptQuirkChanges:]): Deleted.
(-[WKWebPrivacyNotificationListener listenForStorageAccessUserAgentStringQuirkChanges:]): Deleted.
(-[WKWebPrivacyNotificationListener listenForRestrictedOpenerDomainsChanges:]): Deleted.
(WebKit::LinkDecorationFilteringController::shared): Deleted.
(WebKit::LinkDecorationFilteringController::observeUpdates): Deleted.
(WebKit::LinkDecorationFilteringController::setCachedStrings): Deleted.
(WebKit::LinkDecorationFilteringController::updateStrings): Deleted.
(WebKit::StorageAccessPromptQuirkController::shared): Deleted.
(WebKit::StorageAccessPromptQuirkController::observeUpdates): Deleted.
(WebKit::StorageAccessPromptQuirkController::setCachedQuirks): Deleted.
(WebKit::StorageAccessPromptQuirkController::setCachedQuirksForTesting): Deleted.
(WebKit::StorageAccessPromptQuirkController::initialize): Deleted.
(WebKit::StorageAccessPromptQuirkController::updateQuirks): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::shared): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::observeUpdates): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirks): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::setCachedQuirksForTesting): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::initialize): Deleted.
(WebKit::StorageAccessUserAgentStringQuirkController::updateQuirks): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
(WebKit::WebProcessPool::platformInitializeNetworkProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
(WebKit::NetworkProcessProxy::NetworkProcessProxy):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::sendCachedLinkDecorationFilteringData):
(WebKit::WebPageProxy::waitForInitialLinkDecorationFilteringData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setStorageAccessPromptQuirkForTesting):
(WebKit::WebsiteDataStore::setUserAgentStringQuirkForTesting):

Canonical link: https://commits.webkit.org/283464@main
